src/Microsoft.AspNetCore.Rewrite/Internal/UrlActions/CustomResponseAction.cs

```diff
@@ -12,36 +11,31 @@ namespace Microsoft.AspNetCore.Rewrite.Internal.UrlActions
     public class CustomResponseAction : UrlAction
     {
         public int StatusCode { get; }
-        public string StatusReason { get; }
-        public string StatusDescription { get; }
+        public string StatusReason { get; set; }
+        public string StatusDescription { get; set; }
 
-        public CustomResponseAction(int statusCode, string statusReason, string statusDescription)
+        public CustomResponseAction(int statusCode)
         {
-            if (string.IsNullOrEmpty(statusReason))
-            {
-                throw new ArgumentException(nameof(statusReason));
-            }
-
-            if (string.IsNullOrEmpty(statusDescription))
-            {
-                throw new ArgumentException(nameof(statusDescription));
-            }
-
             StatusCode = statusCode;
-            StatusReason = statusReason;
-            StatusDescription = statusDescription;
         }
 
         public override void ApplyAction(RewriteContext context, BackReferenceCollection ruleBackReferences, BackReferenceCollection conditionBackReferences)
         {
-            context.HttpContext.Response.StatusCode = StatusCode;
+            var response = context.HttpContext.Response;
+            response.StatusCode = StatusCode;
 
-            context.HttpContext.Features.Get<IHttpResponseFeature>().ReasonPhrase = StatusReason;
+            if (!string.IsNullOrEmpty(StatusReason))
+            {
+                context.HttpContext.Features.Get<IHttpResponseFeature>().ReasonPhrase = StatusReason;
+            }
 
-            var content = Encoding.UTF8.GetBytes(StatusDescription);
-            context.HttpContext.Response.ContentLength = content.Length;
-            context.HttpContext.Response.ContentType = "text/plain; charset=utf-8";
-            context.HttpContext.Response.Body.Write(content, 0, content.Length);
+            if (!string.IsNullOrEmpty(StatusDescription))
+            {
+                var content = Encoding.UTF8.GetBytes(StatusDescription);
+                response.ContentLength = content.Length;
+                response.ContentType = "text/plain; charset=utf-8";
+                response.Body.Write(content, 0, content.Length);
```

Comment:
The entire middleware and rule infrastructure is sync. This is the only action I've seen where async could even be relevant. I don't think it's worth forking the whole infrastructure fore.